Sources/PackageModel/PackageIdentity.swift

```diff
@@ -94,6 +94,165 @@ extension PackageIdentity: JSONMappable, JSONSerializable {
 
 // MARK: -
 
+extension PackageIdentity {
+    /// Provides a namespace for related packages within a package registry.
+    public struct Scope: LosslessStringConvertible, Hashable, Equatable, Comparable, ExpressibleByStringLiteral {
+        public let description: String
+
+        public init(validating description: String) throws {
```

Comment:
Can we use regex patterns to do the validation?

Reply:
We could, but I can think of a few drawbacks to that approach:
- Less precise errors (a regex match is all or nothing, so you can't know why it's invalid)
- Somewhat slower performance (a regex engine is slower than what we're doing here)
- More coupling (it creates a dependency on a regex engine, which could complicate cross-platform deployment)

On the plus side, a regex is a more obvious representation than string manipulation code. But with comprehensive testing, that's probably less of a concern.